### DISCUSSION TITLE
Bundling multiple notifications into one

### DISCUSSION THREAD

**Discussion by Nixellion** [pymedusa/Medusa#9597] [Ideas & Feature requests]
Hello,

Currently whenever a new episode of the show is found, starts downloading or finishes download Medusa sends one notification per episode in quick succession. Meaning that you can get like 30+ notifications one after another, basically making your PC or phone to go "DING DING DING DING DING!!..." and then on Windows it takes another couple minutes for all notification popups to show up one after another, as it has to play the animate-in and animate-out animations for each popup. At least with slack notifications.

This is bad for multiple reasons. Starting from obviously being annoying and plain distracting if you're working, to going to absolutely ruining message limit quotas on services that have them, like Pushbullet.

**Proposal:**

Make Medusa bundle multiple notifications into a single message. The simplest way I can think of would be making a 'queue' file or in-memory cache, that fills up with notifications and then some kind of separate 'notifier' thread that checks this queue every 5-10 seconds and if it finds notifications it bundles them and sends as one message and then clears the queue or removes sent notifications from the queue. This is the simplest way to implement it that I could think of. There can still be a separate 'immediate notification' function, but at the moment I can't really think of anything Medusa does that requires notifications to come exactly at the moment that 'something' happened. It can be delayed by 5-10 seconds, even a minute and no one would likely care.

Thanks for your attention.

**Comment by p0psicles** [comment]
How would we handle notification balloons that have a char limit?

**Reply by Nixellion** [reply]
Good question. In terms of character limit per message, not balloon - I guess the most straightforward option would be to write down max char limits for every notification platform, and then chunk messages based on that limit. While this may still result in multiple notifications coming one after another, at least it will be a lot less, like 3-5 instead of 30.

As an example, here are my notifications that all came one after another within like 5-10 seconds. That's 38 notifications, with a total of 4392 characters. Slack can handle that as 1 message. Discord has a 2000 character limit and would handle it in 3 messages. Not sure about Pushbullet, but I think it has a long limit as well. 

For extreme cases maybe there could be an option of making it like "Started Download: Better Call Saul - 2x07 and 37 more", which could be optionally enabled for notification agents that don't support long messages.

```
Started Download: Better Call Saul - 2x07 - Inflatable - SDTV with 6 seeders and 1 leechers from NoName Club
Started Download: Better Call Saul - 2x04 - Gloves Off - SDTV with 6 seeders and 1 leechers from NoName Club
Started Download: Better Call Saul - 2x06 - Bali Ha'i - SDTV with 6 seeders and 1 leechers from NoName Club
Started Download: Better Call Saul - 2x03 - Amarillo - SDTV with 6 seeders and 1 leechers from NoName Club
Started Download: Better Call Saul - 2x09 - Nailed - SDTV with 6 seeders and 1 leechers from NoName Club
Started Download: Better Call Saul - 2x02 - Cobbler - SDTV with 6 seeders and 1 leechers from NoName Club
Started Download: Better Call Saul - 2x01 - Switch - SDTV with 6 seeders and 1 leechers from NoName Club
Started Download: Better Call Saul - 2x10 - Klick - SDTV with 6 seeders and 1 leechers from NoName Club
Started Download: Better Call Saul - 3x01 - Mabel - 4K UHD WEB-DL with 3 seeders and 1 leechers from Jackett PB
Started Download: Better Call Saul - 3x02 - Witness - 4K UHD WEB-DL with 3 seeders and 1 leechers from Jackett PB
Started Download: Better Call Saul - 3x03 - Sunk Costs - 4K UHD WEB-DL with 3 seeders and 1 leechers from Jackett PB
Started Download: Better Call Saul - 3x04 - Sabrosito - 4K UHD WEB-DL with 3 seeders and 1 leechers from Jackett PB
Started Download: Better Call Saul - 3x05 - Chicanery - 4K UHD WEB-DL with 3 seeders and 1 leechers from Jackett PB
Started Download: Better Call Saul - 3x06 - Off Brand - 4K UHD WEB-DL with 3 seeders and 1 leechers from Jackett PB
Started Download: Better Call Saul - 3x07 - Expenses - 4K UHD WEB-DL with 3 seeders and 1 leechers from Jackett PB
Started Download: Better Call Saul - 3x08 - Slip - 4K UHD WEB-DL with 3 seeders and 1 leechers from Jackett PB
Started Download: Better Call Saul - 3x09 - Fall - 4K UHD WEB-DL with 3 seeders and 1 leechers from Jackett PB
Started Download: Better Call Saul - 3x10 - Lantern - 4K UHD WEB-DL with 3 seeders and 1 leechers from Jackett PB
Started Download: Better Call Saul - 4x01 - Smoke - 1080p BluRay with 8 seeders and 11 leechers from Jackett PB
Started Download: Better Call Saul - 4x02 - Breathe - 1080p BluRay with 8 seeders and 11 leechers from Jackett PB
Started Download: Better Call Saul - 4x03 - Something Beautiful - 1080p BluRay with 8 seeders and 11 leechers from Jackett PB
Started Download: Better Call Saul - 4x04 - Talk - 1080p BluRay with 8 seeders and 11 leechers from Jackett PB
Started Download: Better Call Saul - 4x05 - Quite a Ride - 1080p BluRay with 8 seeders and 11 leechers from Jackett PB
Started Download: Better Call Saul - 4x06 - Piñata - 1080p BluRay with 8 seeders and 11 leechers from Jackett PB
Started Download: Better Call Saul - 4x07 - Something Stupid - 1080p BluRay with 8 seeders and 11 leechers from Jackett PB
Started Download: Better Call Saul - 4x08 - Coushatta - 1080p BluRay with 8 seeders and 11 leechers from Jackett PB
Started Download: Better Call Saul - 4x09 - Wiedersehen - 1080p BluRay with 8 seeders and 11 leechers from Jackett PB
Started Download: Better Call Saul - 4x10 - Winner - 1080p BluRay with 8 seeders and 11 leechers from Jackett PB
Started Download: Better Call Saul - 5x01 - Magic Man - 720p WEB-DL with 121 seeders and 15 leechers from Jackett PB
Started Download: Better Call Saul - 5x02 - 50% Off - 720p WEB-DL with 121 seeders and 15 leechers from Jackett PB
Started Download: Better Call Saul - 5x03 - The Guy for This - 720p WEB-DL with 121 seeders and 15 leechers from Jackett PB
Started Download: Better Call Saul - 5x04 - Namaste - 720p WEB-DL with 121 seeders and 15 leechers from Jackett PB
Started Download: Better Call Saul - 5x05 - Dedicado a Max - 720p WEB-DL with 121 seeders and 15 leechers from Jackett PB
Started Download: Better Call Saul - 5x06 - Wexler v. Goodman - 720p WEB-DL with 121 seeders and 15 leechers from Jackett PB
Started Download: Better Call Saul - 5x07 - JMM - 720p WEB-DL with 121 seeders and 15 leechers from Jackett PB
Started Download: Better Call Saul - 5x08 - Bagman - 720p WEB-DL with 121 seeders and 15 leechers from Jackett PB
Started Download: Better Call Saul - 5x09 - Bad Choice Road - 720p WEB-DL with 121 seeders and 15 leechers from Jackett PB
Started Download: Better Call Saul - 5x10 - Something Unforgivable - 720p WEB-DL with 121 seeders and 15 leechers from Jackett PB
```

---

For balloons - well, I think they are handled by OS already. I mean here's an example notification:

`Started Download: Better Call Saul - 3x03 - Sunk Costs - 4K UHD WEB-DL with 3 seeders and 1 leechers from Jackett PB`

It's already longer than any balloon can handle. I almost never see entire notification text on Windows or Android, it's always trimmed like "Started Download: Better Call Saul - 3x03 - Sunk Co..." or something like that. So I don't think it's something that Medusa is supposed to handle on it's end?


PS:
No, I did not just add this show, for some reason it's been sitting for months and only now it found all that. Maybe someone uploaded it I dunno :)

**Reply by purrloftruth** [reply]
> How would we handle notification balloons that have a char limit?

I don't think that downloads of multiple episodes necessarily need to list every episode.

In situations where, for example, a season pack torrent is downloaded and, as such, the torrent contains multiple episodes, simply notifying about the download of *the torrent* instead of each individual episode is more reasonable:

`Started Download: TV Show - Season 1 - 4K UHD WEB-DL with 194 seeders and 5 leechers from Client`

in situations where multiple episodes of a single show are being downloaded, but not in the form of a season pack, they could potentially be bundled as such:

`Started Download: TV Show - 1x01, 1x04, 1x05, (3 more episodes...) - 4K UHD WEB-DL from Client`

Notice that the seeder/leecher counts would be elided in this situation.

When a download finished notification needs to be sent for an episode in that grouping of episodes, a timer of 1 minute could be started, and after 1 minute all episodes that finished downloading in that time could be notified in a single notification:

`Download Finished: TV Show - 1x01, 1x05 - 4K UHD WEB-DL`

If all given episodes in the grouping finished within the span of that minute, it would cancel the timer and send the bundled notification immediately.

[The way that Medusa currently handles notifications for multiple episodes is obnoxious.](https://bsky.app/profile/purrl.org/post/3m7sk3f2cl22m) I've been using Medusa for years and years, but I've gotten fed up with this behavior, and I'm currently strongly considering moving to Sonarr _simply to resolve this issue_. I think, regardless of whether I do or not, this is a relatively important improvement to make, as you can't ever know the number of people who are frustrated by this yet not coming to GitHub to complain about it.

Thanks for your constant hard work.
